events: deprecate static listenerCount function

thefourtheye opened the pull request:
As per the discussion in #734, this patch deprecates the usage of
`EventEmitter.listenerCount` static function and introduces the `listenerCount`
function in the prototype of `EventEmitter` itself.

mscdex commented on lib/events.js:
```diff
@@ -394,7 +396,28 @@ EventEmitter.prototype.listeners = function listeners(type) {
+
+  const events = this._events;
+
+  if (events && events.hasOwnProperty(type)) {
```
Wouldn't it be better to use something like `Object.prototype.hasOwnProperty.call(events, type)` instead?

Also FWIW `hasOwnProperty()` still causes a significant performance hit and would technically not be backwards compatible behavior.

thefourtheye replied:
Oh yeah, I'll remove that. Anyway, if #2350 lands, this will not matter.

mscdex commented on lib/events.js:
```diff
@@ -394,7 +396,28 @@ EventEmitter.prototype.listeners = function listeners(type) {
+EventEmitter.prototype.listenerCount = function listenerCount(type) {
+
+  if (typeof type !== 'string') {
+    throw new TypeError('type must be a valid string');
```
The wording here may be a bit misleading since we don't actually validate `type` when it *is* a string. Maybe just say `type must be a string`?

thefourtheye replied:
@mscdex Done!

targos commented:
FYI there was some discussion in https://github.com/nodejs/dev-policy/issues/49 about the deprecation steps and it seems most people agree that we should start only with documentation. The discussion is quite old now so correct me if I missed something else.

brendanashworth commented:
A few thoughts:
- Probably needs to wait for https://github.com/nodejs/dev-policy/issues/76 to be determed a minor or major change
- Needs other uses in the code base to be changed away from `EventEmitter.listenerCount(...)`, so it doesn't emit deprecation notices
- Deprecation policy (`EventEmitter.listenerCount` might be used in too many places already)

thefourtheye commented:
@brendanashworth @targos I removed the deprecation from the code.

@brendanashworth I updated other parts of the code which use `EventEmitter.listenerCount` to the prototype's `listenerCount`

thefourtheye commented:
I included a benchmark for the prototype's `listenerCount` as well and the results are as follows

    ➜  io.js git:(event-emitter-listener-count) ./iojs benchmark/events/ee-listener-count.js 
    events/ee-listener-count.js n=50000000: 241760077.96588
    ➜  io.js git:(event-emitter-listener-count) ./iojs benchmark/events/ee-listener-count-on-prototype.js 
    events/ee-listener-count-on-prototype.js n=50000000: 286204676.98511

ChALkeR commented:
Personally, I'm +1 on this (with a documentation-only deprecation for now).

I guess you should wait for @trevnorris response.

See also https://github.com/nodejs/node/issues/1817#issuecomment-106130974 and https://github.com/nodejs/node/issues/734#issuecomment-76272058.

trevnorris commented on doc/api/events.markdown:
```diff
@@ -137,10 +137,15 @@ Execute each of the listeners in order with the supplied arguments.
 
+### emitter.listenerCount(type)
+
+Returns the number of listeners listening to the `type` of event.
```
Since we're making a change, add a bullet that `type` is a `{String}`. As done in `'removeListener'`.

trevnorris commented on doc/api/events.markdown:
```diff
@@ -137,10 +137,15 @@ Execute each of the listeners in order with the supplied arguments.
 
 Return the number of listeners for a given event.
 
+_Note: This is deprecated. Use <EventEmitter object>.listenerCount instead._
```
Don't use angle brackets. The combination of spacing and italics is causing the words to be removed.

trevnorris commented on lib/events.js:
```diff
@@ -394,6 +394,27 @@ EventEmitter.prototype.listeners = function listeners(type) {
 
+EventEmitter.prototype.listenerCount = function listenerCount(type) {
+
+  if (typeof type !== 'string') {
```
remove whitespace.

thefourtheye replied:
I am not sure which one to remove. Can you please be specific?

trevnorris replied:
Between top of function and first `if`.

thefourtheye replied:
Done

trevnorris commented on lib/events.js:
```diff
@@ -394,6 +394,27 @@ EventEmitter.prototype.listeners = function listeners(type) {
+  }
+
+  return 0;
+};
```
Make `EventEmitter.listenerCount()` simply `return emitter.listenerCount(type)`. Prevent unlikely but possible divergence between the two.

thefourtheye replied:
@trevnorris Ah, then I need to remove the type checking from this function. Otherwise it might break existing code :-/ Shall I remove that?

trevnorris replied:
Missed that. Should we be checking type? It's a slight change that deviates from previous impl, and so we can't unequivocally say it's possible to just switch usage. I don't care either way. If the check stays in the yeah, can leave it just as is.

thefourtheye replied:
Cool, I am leaving the check then.

targos replied:
We shouldn't enforce string for the type. It would not be possible to count events using a Symbol.
Furthermore, if we eventually switch to a Map implementation, any type could be used.

trevnorris replied:
Oop, yup. Remove the string check:
```js
'use strict';

const EE = require('events');
const util = require('util');

let e = new EE();

let b = {};

e.on(b, function() { console.log('hi'); });
e.emit(b);  // hi
```
Thanks @targos

trevnorris commented on test/parallel/test-event-emitter-listener-count.js:
```diff
@@ -0,0 +1,15 @@
+
+const common = require('../common');
+const assert = require('assert');
+const EventEmitter = require('events').EventEmitter;
```
Don't need to include the `.EventEmitter` at the end.

trevnorris commented:
Left a couple comments, but looks good.

thefourtheye commented:
@trevnorris I addressed the comments, PTAL now. I removed the `benchmark/events/ee-listener-count.js` as it is just a wrapper for the `listenerCount` on the prototype.

thefourtheye commented:
Rebased with master and squashed the commits. PTAL.

trevnorris commented on doc/api/events.markdown:
```diff
@@ -137,10 +137,17 @@ Execute each of the listeners in order with the supplied arguments.
 
+### emitter.listenerCount(type)
+
+* `type` {String} The type of event
```
Change `{String}` to `{Value}`

thefourtheye replied:
But we have a check now, which expects the `type` to be string. Is that okay to change this to `Value`?

trevnorris replied:
The check should've been removed: https://github.com/nodejs/node/pull/2349#discussion_r37452127

thefourtheye replied:
Oops, sorry. There is no clear way to know if there are multiple updates in the same PR. I missed that comment :( I'll update it now.

trevnorris replied:
No worries. Somehow it was placed as a comment in old code segment. Had me confused for a bit.

thefourtheye replied:
Done! PTAL now.

trevnorris commented:
LGTM

thefourtheye commented:
@targos @brendanashworth @ChALkeR @mscdex PTAL and let me know if it LGTY

mscdex commented on doc/api/events.markdown:
```diff
@@ -137,10 +137,17 @@ Execute each of the listeners in order with the supplied arguments.
 
 Return the number of listeners for a given event.
 
+_Note: This is deprecated. Use `eventEmitterObject.listenerCount` instead._
```
Maybe `eventEmitterObject` should just be `emitter` to match the style of the documented instance properties.

thefourtheye replied:
Thanks :-) Fixed.

mscdex commented:
One minor doc nit, but other than that LGTM.

thefourtheye commented:
This should be landed in `master` only or `v3.x` as well?

trevnorris commented:
@thefourtheye Only worry about `master` for now. Can be cherry-picked later before next release (doubt this would make the 3.1.0 release).

thefourtheye commented:
@trevnorris Thanks :-)

Landed at 8f58fb92fff904a6ca58fd0df9ee5a1816e5b84e. Thanks people :-)
